src/main/java/vazkii/botania/data/RecipeAdvancementProvider.java

```diff
@@ -0,0 +1,260 @@
+				if (triggerIngredients.length > 1 || triggerIngredients.length == 0) {
+					LOGGER.info(entry.getValue().getRecipeOutput().getItem().getRegistryName());
+					for (Object o : triggerIngredients) {
+						LOGGER.info("    " + ((Ingredient) o).serialize());
+					}
+					continue;
+				} else {
+					trigger = (Ingredient) triggerIngredients[0];
+				}
+			}
+			JsonObject advancement = Advancement.Builder.builder().withParentId(parentAdvancement)
+					.withCriterion("has_the_recipe", new RecipeUnlockedTrigger.Instance(entry.getKey()))
+					.withCriterion("has_required_item", hasItem(Items.COBBLESTONE))
+					.withRewards(AdvancementRewards.Builder.recipe(entry.getKey()))
+					.withRequirementsStrategy(IRequirementsStrategy.OR).serialize();
+			JsonArray items = advancement.getAsJsonObject("criteria").getAsJsonObject("has_required_item")
+					.getAsJsonObject("conditions").getAsJsonArray("items");
+			items.remove(0);
+			items.add(trigger.serialize());
+			saveRecipeAdvancement(cache, advancement,
+					generator.getOutputFolder()
+							.resolve("data/" + entry.getValue().getId().getNamespace() + "/advancements/recipes/"
+									+ getItemCategory(entry.getValue().getRecipeOutput().getItem()) + "/"
+									+ entry.getValue().getId().getPath() + ".json"));
+		}
+	}
+
+	@Override
+	public String getName() {
+		return "Botania recipe advancements";
+	}
+
+	private RecipeManager loadRecipesAndTags() {
+		ResourcePackList<ResourcePackInfo> resourcePacks = new ResourcePackList<>(ResourcePackInfo::new);
+		ResourcePackLoader.loadResourcePacks(resourcePacks, RecipeAdvancementProvider::buildPackFinder);
+		SimpleReloadableResourceManager resourceManager = new SimpleReloadableResourceManager(
+				ResourcePackType.SERVER_DATA, Thread.currentThread());
+		resourceManager.addReloadListener(new NetworkTagManager());
+		RecipeManager recipeManager = new RecipeManager();
+		resourceManager.addReloadListener(recipeManager);
+		resourcePacks.addPackFinder(new ServerPackFinder());
+		resourcePacks.reloadPacksFromFinders();
+		List<ResourcePackInfo> list = Lists.newArrayList(resourcePacks.getEnabledPacks());
+		for (ResourcePackInfo resourcepackinfo : resourcePacks.getAllPacks()) {
+			resourcepackinfo.getPriority().func_198993_a(list, resourcepackinfo, (p_200247_0_) -> {
+				return p_200247_0_;
+			}, false);
+		}
+		resourcePacks.setEnabledPacks(list);
+		List<IResourcePack> list1 = Lists.newArrayList();
+		resourcePacks.getEnabledPacks().forEach((p_200244_1_) -> {
+			list1.add(p_200244_1_.getResourcePack());
+		});
+		CompletableFuture<Unit> completablefuture = resourceManager.reloadResourcesAndThen(Util.getServerExecutor(),
+				Util.getServerExecutor(), list1, CompletableFuture.completedFuture(Unit.INSTANCE));
+		BooleanSupplier isDone = completablefuture::isDone;
+		while (!isDone.getAsBoolean()) {
+			Thread.yield();
+			LockSupport.parkNanos("waiting for tasks", 100000L);
+		}
+		return recipeManager;
```

Comment:
is this actually required to have recipe advancements? 0.o
seems extremely hacky

Reply:
the recipes need to be loaded for the Data Generator to load them and this is as close to the Vanilla Way as i could get.

Reply:
it should work without the Thread.yield if that is what you mean,
but without anything to do inside the while it somehow just freezes forever.

Reply:
It would also work with a
```java
try {
    Thread.sleep(10);
} catch(Exception e) {
}
`
inside of the loop, but i am not sure thats a better solution.

Reply:
This loop could also be replaced with `completablefuture.join();`.